`unused_mut` false positive due to borrow-checking error

### Code

```Rust
struct Thing;
impl Drop for Thing {
    fn drop(&mut self) {}
}

fn main() {
    let mut t;
    let mut b = None;
    loop {
        t = Thing;
        b.insert(&t);
    }
}
```

### Current output

```Shell
warning: variable does not need to be mutable
 --> src/main.rs:7:9
  |
7 |     let mut t;
  |         ----^
  |         |
  |         help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` (part of `#[warn(unused)]`) on by default

error[E0506]: cannot assign to `t` because it is borrowed
  --> src/main.rs:10:9
   |
10 |         t = Thing;
   |         ^ `t` is assigned to here but it was already borrowed
11 |         b.insert(&t);
   |         -        -- `t` is borrowed here
   |         |
   |         borrow later used here

For more information about this error, try `rustc --explain E0506`.
warning: `playground` (bin "playground") generated 1 warning
error: could not compile `playground` (bin "playground") due to 1 previous error; 1 warning emitted
```

### Desired output

```Shell
The unused_mut lint should not fire
```

### Rationale and extra context

_No response_

### Other cases

```Rust

```

### Rust Version

```Shell
Reproducible on the playground with version 1.95.0-nightly (2026-02-01 57d2fb136650d05efb3e)
```

### Anything else?

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_borrowck/src/lib.rs
+++ b/compiler/rustc_borrowck/src/lib.rs
@@ -1207,6 +1207,17 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, '_, 'tcx> {
                 "access_place: suppressing error place_span=`{:?}` kind=`{:?}`",
                 place_span, kind
             );
+
+            // If the place is being mutated, then mark it as such anyway in order to suppress the
+            // `unused_mut` lint, which is likely incorrect once the access place error has been
+            // resolved.
+            if rw == ReadOrWrite::Write(WriteKind::Mutate)
+                && let Ok(root_place) =
+                    self.is_mutable(place_span.0.as_ref(), is_local_mutation_allowed)
+            {
+                self.add_used_mut(root_place, state);
+            }
+
             return;
         }
 

--- a/tests/ui/lint/unused/mut-used-despite-borrowck-error.rs
+++ b/tests/ui/lint/unused/mut-used-despite-borrowck-error.rs
@@ -1,0 +1,19 @@
+//! Do not fire unused_mut lint when mutation of the bound variable fails due to a borrow-checking
+//! error.
+//!
+//! Regression test for https://github.com/rust-lang/rust/issues/152024
+//@ compile-flags: -W unused_mut
+
+struct Thing;
+impl Drop for Thing {
+    fn drop(&mut self) {}
+}
+
+fn main() {
+    let mut t;
+    let mut b = None;
+    loop {
+        t = Thing; //~ ERROR cannot assign to `t` because it is borrowed
+        b.insert(&t);
+    }
+}

--- a/tests/ui/lint/unused/mut-used-despite-borrowck-error.stderr
+++ b/tests/ui/lint/unused/mut-used-despite-borrowck-error.stderr
@@ -1,0 +1,13 @@
+error[E0506]: cannot assign to `t` because it is borrowed
+  --> $DIR/mut-used-despite-borrowck-error.rs:16:9
+   |
+LL |         t = Thing;
+   |         ^ `t` is assigned to here but it was already borrowed
+LL |         b.insert(&t);
+   |         -        -- `t` is borrowed here
+   |         |
+   |         borrow later used here
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0506`.

